ExodusMovement/schemasafe
# ci: automatic releases

Automates releases by releasing on merge to master for relevant commit types (`feat`, `fix`, `perf`, and breaking changes).

~If we want to use `@semantic-release/git`, which we should, we'll have to use a PAT of the machine account and empower that machine account to bypass branch protection rules. The plugin is responsible for updating the version in `package.json` and adding the `CHANGELOG.md`!~ Decided to use `@semantic-release/github` instead for tighter security so we don't have to allow the machine user to bypass the branch protection rule. This will result in stale package versions in git, but still have the correct version in the published package. 

## Todos:
- [x] create a `NPM_PUBLISH_TOKEN` secret that has an NPM automation token for the account that will get exclusive access to `@exodus/schemasafe`

## Testplan
A first sanity check of the configuration can be done using the dry run mode locally.
1. Checkout this PR
2. Apply the following diff

```diff
Subject: [PATCH] add test branch
---
Index: .releaserc.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.releaserc.json b/.releaserc.json
--- a/.releaserc.json	(revision f9059cf10c831fad0f2a0fc8d17ab1d1c6fc8a85)
+++ b/.releaserc.json	(date 1699578894912)
@@ -1,5 +1,5 @@
 {
-  "branches": ["master"],
+  "branches": ["master", "sparten11740/ci/automatic-releases"],
   "plugins": [
     [
       "@semantic-release/commit-analyzer",

```
3. run `yarn release -d`. verify `echo $?` returns `0` and you see some nice release notes in the stdout of the first command

This doesn't test the full integration such as correctly configured PAT's and branch protection rules. We'll only be able to test these when this PR gets merged
